### PR TITLE
driver: display: Add support for BBC Micro:bit v2 to display driver

### PR DESCRIPTION
**Pull request by chudov**
Yet another attempt to add support of display driver for BBC Micro:bit v2 board.

This PR uses Devicetree for pin assignment. Existing code is reused as much as possible. 

~~Unfortunately driver initialization code looks ugly, as I don't yet fully understand how all these DT macros can be used for initialization.  Can someone consult me on it?~~

Tested on BBC Micro:bit v2, but not on Micro:bit v1.

Comments for further improvements and test results on v1 board are very welcome.

Signed-off-by: Alexander Chudov <chudov@gmail.com>

### CODE REVIEW

**Comment by peterniebert** [comment]
Dear Alexander,
I have to admit that I gave up on my own contribution this summer because the processing of the PR did not go together with my own timing and I had to think about other things. This autumn, I did some practical exercises about this with my students that kind of go in your way, so here is a proposition for improvement. Or even to work together to get things right.
I am fed up with the poor v2 implementation level, a shame for such a great development card.

So please read me and give me your thoughts. 

But first things first, DTS "fun" (DTS is the hardest part of Zephyr, I think).

For one, the matrix can be defined simpler in DTS using actual arrays, like so for the v2 (sorry for the indenting that looks strange here with tabs):
```
	matrix {

        compatible = "gpio-leds";

        rows {
				gpios=
					<&gpio0 21 GPIO_ACTIVE_HIGH>,
                	<&gpio0 22 GPIO_ACTIVE_HIGH>,
                	<&gpio0 15 GPIO_ACTIVE_HIGH>,
                	<&gpio0 24 GPIO_ACTIVE_HIGH>,
                	<&gpio0 19 GPIO_ACTIVE_HIGH>;

                label = "Matrix rows";

        };

        cols {

            gpios=
					<&gpio0 28 GPIO_ACTIVE_LOW>,
                	<&gpio0 11 GPIO_ACTIVE_LOW>,
                	<&gpio0 31 GPIO_ACTIVE_LOW>,
                	<&gpio1 5 GPIO_ACTIVE_LOW>,
                	<&gpio0 30 GPIO_ACTIVE_LOW>;
            label = "Matrix colums";

        };

    };
```

Then, you can define const arrays for the indexed access to each entry, like so (here obviously not parametrized for an alternative driver):

```

#include <zephyr.h>
#include <device.h>
#include <devicetree.h>
#include <drivers/gpio.h>
#include <devicetree/gpio.h>

#include "led_matrix.h"

const int led_matrix_rows = 5;
const int led_matrix_columns = 5;

// tables for dynamic access to pins and devices

const struct gpio_dt_spec  row_pin[5] = {
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, rows), gpios, 0),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, rows), gpios, 1),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, rows), gpios, 2),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, rows), gpios, 3),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, rows), gpios, 4)
};

const struct gpio_dt_spec col_pin[5] = {
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, cols), gpios, 0),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, cols), gpios, 1),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, cols), gpios, 2),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, cols), gpios, 3),
	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(matrix, cols), gpios, 4),
};


void led_matrix_init(void)
{
	for (int i = 0; i < 5; i++) {



		gpio_pin_configure_dt(&row_pin[i], GPIO_OUTPUT);
		gpio_pin_configure_dt(&col_pin[i], GPIO_OUTPUT);
	}

}

void led_matrix_clear(void)
{
		for (int i = 0; i < 5; i++) {
			gpio_pin_set_dt(&row_pin[i], 0);
			gpio_pin_set_dt(&col_pin[i], 0);

	}
}
``` 

Students managed to go from here to code from zero in a short time a display driver thread that would sweep over the matrix and let some time (1ms) for other activity at each pixel, and the result looks OKish.

However, the whole thing remains terribly limited and inefficient, in particular compared to the driver present in CodeMaker software (based on NRF5SDK), for instance, that driver allows for changing brightness.

Moreover, the original author of the Zephyr driver tried to be smart and use line multiplexing (showing several pixels in the same row or same column). If you think about it, that is a cause of trouble in one of two ways: Either several LEDs take their current through the same resistance, but then, the more leds you switch on, the less bright they will be, leading to an inhomogenous brightness. Or the current of all LEDs adds up and passes through a single GPIO pin exceeding the specs of that pin, and this might over time damage the pin. 

That is neither the case for the CodeMaker driver, not for my own attempt in Zephyr directly using NRFX this summer. It is better to sweep over pins individually. But then, the code has to be efficient.


Using the gpio_pin_set_dt function obviously is not efficient as it inserts an actual function call in pin manipulation, although this is perfectly avoidable with the NRFX API. Secondly, the NRFX API allow to increase current limitations of the GPIO pins and that allows to obtain a brighter display when sweeping over LEDs individually.

Moreover, using Kernel scheduling is not efficient for a device driver like this. Rather, a direct interrupt timer approach puts much less load on the system.

And in the end I think, that the reuse of the code of the original driver is not such a good idea as even the API is bad. I think a from scratch development that also implements in the API brighness control, and the the brightness sensor feature would be desirable. And then, a modularisation between the basic matrix display and the contents API (text, images) is desirable. The contents can and should totally be done with Zephyr means, whereas the basic display driver should be implemented low level to be efficient. 

In other PRs the discussion also turned around the topic of this modularity as zephyr already has matrix displays, but the microbit display is so tiny (5x5) that it is reasonable to give it a separate contents API.

Thanks for reading this far, please give me your thoughts!

**Comment by chudov** [comment]
Dear Peter,

thank you for suggestions for DTS. I'm trying to use the proposed approach... DTS, and especially all these macros are really tricky to understand :)

Loose of brightness - yes, possible. To be checked how big might be the impact. From another perspective - it is not a safety-relevant device, some small changes in brightness should not be a big issue.

It seems overcurrent is impossible with the current schematic: the current per column is limited to 3 mA. So even if all LEDs are switched on, the maximum sink current over MCU will be 15 mA, source current is 12 mA from gpio0 and 3 mA from gpio1. That is completely in allowed range.

I don't think direct usage of interrupt controller is a good idea as at the very end there is a resource consumer outside of the operating system that steals resources from it. And there is no way to control this consumer explicitly from the operating system.

File: boards/arm/bbc_microbit/bbc_microbit.dts [hunk 1]
```diff
@@ -40,6 +41,46 @@
...
+		};
+		row_3: led_row3 {
+			gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;
+		};
+
+		col_1: led_col1 {
+			gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
+		};
+		col_2: led_col2 {
+			gpios = <&gpio0 5 GPIO_ACTIVE_LOW>;
```
Comment (gmarull) [inline boards/arm/bbc_microbit/bbc_microbit.dts:60]: suggestion: create a specific compatible and create something like:

```
led_matrix {
    compatible = "bbc,microbit-matrixled";
    row-gpios = <&gpio0 13 GPIO_ACTIVE_HIGH>, <&gpio0 14 GPIO_ACTIVE_HIGH>, ...;
    col-gpios = <&gpio0 4 GPIO_ACTIVE_LOW>, ...;
```

File: drivers/display/mb_display.c [hunk 2]
```diff
@@ -429,25 +426,43 @@ struct mb_display *mb_display_get(void)
...
-	gpio_pin_configure(display.dev, LED_COL7_GPIO_PIN, GPIO_OUTPUT);
-	gpio_pin_configure(display.dev, LED_COL8_GPIO_PIN, GPIO_OUTPUT);
-	gpio_pin_configure(display.dev, LED_COL9_GPIO_PIN, GPIO_OUTPUT);
-
+#if defined(CONFIG_BOARD_BBC_MICROBIT)
+	display.dev[0] = device_get_binding(DT_LABEL(DT_NODELABEL(gpio0)));
+
+	__ASSERT(dev[0], "No GPIO device found");
+
+	gpio_pin_configure(display.dev[0], LED_PIN(row_1), GPIO_OUTPUT);
```
Comment (gmarull) [inline drivers/display/mb_display.c:434]: please, take the opportunity to make this driver use `gpio_configure_dt` API (and related DT macros/API)

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/arm/bbc_microbit/bbc_microbit.dts (hunk 1 above)
Comment (chudov) [reply boards/arm/bbc_microbit/bbc_microbit.dts:60]: I'm trying to follow the approach. But how then use these row_gpios and col_gpios in a universal way for anything? For MicrobitV1 row-gpios contains 3 elements and col-gpios 9 elements. For MicrobitV2 they are 5 and 5 elements. Ok, with `DT_INST_PROP_LEN(0, row_gpios)` it is possible to know how many elements are in these `phandle-arrays`. But how then this information can be used for pin initialization and for pin settings? 
```
for(uint8_t i=0; i < DT_INST_PROP_LEN(0, row_gpios); i++) {
     GPIO_DT_SPEC_INST_GET_BY_IDX(0, row_gpios,i)
}
``` 

doesn't work.  
Is there any way to iterate over all `phandle-array` elements to generate an initialization structure like
```
static const struct gpio_dt_spec row_gpio_dt_spec[] = {
   {GPIO_DT_SPEC_INST_GET_BY_IDX(0, row-gpios,0)},
   {GPIO_DT_SPEC_INST_GET_BY_IDX(0, row-gpios,1)},
   ...
   {GPIO_DT_SPEC_INST_GET_BY_IDX(0, row-gpios,DT_INST_PROP_LEN(0, row_gpios))},
}; 
```
?

I thought `DT_INST_FOREACH_CHILD` does the magic, but I failed with it.

Would you please help?

Comment (chudov) [reply boards/arm/bbc_microbit/bbc_microbit.dts:60]: I can answer to the questions by myself: it is needed to create configuration for the driver manually. This is a board-specific static array that actually maps logical row and column numbers to these DT elements. The initialization then is done by iteration over elements of this array (actually, arrays).

Comment (chudov) [reply boards/arm/bbc_microbit/bbc_microbit.dts]: Magic fixed, proposal implemented.

File: drivers/display/mb_display.c (hunk 2 above)
Comment (chudov) [reply drivers/display/mb_display.c]: DT is used as much as for the moment possible.

**Comment by chudov** [comment]
Further steps for the driver implementation are:

- [ ] 1. To finalize DT usage for configuration:

> 1. Eliminate `col-gpios-used` - this information is available in `DT_N_S_matrix_REQUIRES_ORDS`, just need to understand how to take and use it
> 1. Find a way to define pixel_map in device tree, not in sources
> 2. Find a way to build display_cfg.mask at build time instead of display.mask that is build at runtime

- [ ] 2. To use standard Display Driver API instead of self-made API from the previous version of the driver

- [ ] 3. To implement support of brightness change and other stuff that is available in the original CODAL driver

- [ ] 4. To find a way to implement light detection. Perhaps it will be another driver - to be investigated.

**Comment by chudov** [comment]
After #40968 this MR make no sense.

**Pull request closed without merging**